packages/react-query/src/utils/structuralSharing.ts

```diff
@@ -0,0 +1,132 @@
+/**
+ * Inspired by @tanstack/query-core utils.ts
+ * Modified to support Date, Map, Set and NaN comparisons
+ *
+ * This function returns `a` if `b` is deeply equal.
+ * If not, it will replace any deeply equal children of `b` with those of `a`.
+ */
+export function createStructuralSharingFunction(
+  equalityFn: (a: any, b: any) => boolean,
+) {
+  const structuralSharingFunction = (a: any, b: any): any => {
+    if (equalityFn(a, b)) {
+      return a;
+    }
+
+    const array = isPlainArray(a) && isPlainArray(b);
+    const plainObject = isPlainObject(a) && isPlainObject(b);
+    const map = a instanceof Map && b instanceof Map;
+    const set = a instanceof Set && b instanceof Set;
+
+    if (array || plainObject || map) {
+      const aItems = array
+        ? a
+        : plainObject
+          ? Object.keys(a)
+          : Array.from(a.keys());
+      const aSize = aItems.length;
+      const bItems = array
+        ? b
+        : plainObject
+          ? Object.keys(b)
+          : Array.from(b.keys());
+      const bSize = bItems.length;
+      const copy: any = array ? [] : plainObject ? {} : new Map();
+
+      let equalItems = 0;
+
+      for (let i = 0; i < bSize; i++) {
+        const key = array ? i : bItems[i];
+        if (
+          !array &&
+          a[key] === undefined &&
+          b[key] === undefined &&
+          aItems.includes(key)
+        ) {
+          copy[key] = undefined;
+          equalItems++;
+        } else {
+          copy[key] = structuralSharingFunction(a[key], b[key]);
+          if (copy[key] === a[key] && a[key] !== undefined) {
+            equalItems++;
+          }
+        }
+      }
+
+      return aSize === bSize && equalItems === aSize ? a : copy;
+    }
+
+    if (set) {
+      const copy: any = new Set();
+      let equalItems = 0;
+      if (a.size !== b.size) return false;
+      const aItems = Array.from(a.values());
+      for (const bItem of b.values()) {
+        // We're doing a shallow comparison here, not a deep one.
+        const aItem = aItems.find((aItem) => equalityFn(aItem, bItem));
+        if (!aItem) {
+          copy.add(bItem);
+        } else {
+          copy.add(aItem);
+          equalItems++;
+        }
+      }
+      return a.size === b.size && equalItems === a.size ? a : copy;
+    }
+
+    return b;
+  };
+
+  return structuralSharingFunction;
+}
+
+export const defaultStructuralSharingFunction =
+  createStructuralSharingFunction(isEqual);
+
+function isEqual(a: any, b: any) {
```

Comment:
```suggestion
export function defaultIsEqual(a: unknown, b: unknown) {
```

export it so it can be used when making a custom `createStructuralSharingFunction`